Model/statemachine.py

```diff
@@ -997,44 +997,63 @@ def cancel_coin_selec_0(self, vault, needed_fee, feerate):
         return coins
 
     def cancel_coin_selec_1(self, vault, needed_fee, feerate):
-        """Select the combination that results in the smallest overpayment"""
+        """Select the combination of fee-bumping coins that results in the
+        smallest overpayment possible.
+
+        The UTxO pool is laid out with large coins covering up to the reserve
+        and smaller coins used for a finer grained coin selection to avoid
+        overpayments.
+        First try to find the number of Vb (large) coins needed to cover for
+        the most part of the fees, then fill the gap with Vm (small) coins.
+        """
         coins = []
-
-        best_combination = None
-        min_fee_added = None
-        max_paying_combination = None
-        max_fee_added = 0
-        allocated_coins = vault.allocated_coins()
-        for candidate in itertools.chain.from_iterable(
-            itertools.combinations(allocated_coins, r)
-            for r in range(1, len(allocated_coins) + 1)
-        ):
-            added_fees = sum(
-                [c.amount - P2WPKH_INPUT_SIZE * feerate for c in candidate]
+        allocated_coins = sorted(vault.allocated_coins(), key=lambda c: c.amount)
+        # All vb coins are always larger than vm coins
+        vm_coins, vb_coins = (
+            allocated_coins[: -self.vb_coins_count],
+            allocated_coins[-self.vb_coins_count :],
+        )
+        # All vb coins are created equal
+        vb = vb_coins[0].amount
+        txin_cost = P2WPKH_INPUT_SIZE * feerate
+
+        def coin_sum(coins):
+            return sum(c.amount for c in coins)
+
+        # First check if the needed amount is very low, in which case we don't
+        # even need a Vb coin.
+        if vb_coins[0].amount > needed_fee + txin_cost:
+            for i in range(1, len(vm_coins)):
+                if coin_sum(vm_coins[:i]) >= needed_fee + txin_cost * i:
+                    return vm_coins[:i]
+            return [vb_coins[0]]
+
+        # Then gather enough vb coins
+        paid = 0
+        for i in range(1, len(vb_coins)):
+            if needed_fee - paid < vb - txin_cost:
+                break
+            coins.append(vb_coins[i])
+            paid += vb - txin_cost
+
+        # And finally fill the gap with small Vm coins. Note we go through the
+        # list in reverse order as the Vm coins amount is increasing.
+        for i in range(len(vm_coins), i):
```

Comment:
```python
for i in range(len(vm_coins), -1, -1)):
```

Reply:
Woops good catch!

Reply:
Shouldn't end at `-1` tho :p

Reply:
`range()` doesn't include the final digit